cirq-core/cirq/transformers/noise_adding.py

```diff
@@ -0,0 +1,112 @@
+# Copyright 2024 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections.abc import Mapping
+
+from cirq import ops, circuits
+from cirq.transformers import transformer_api
+import numpy as np
+
+
+def _gate_in_moment(gate: ops.Gate, moment: circuits.Moment) -> bool:
+    """Check whether `gate` is in `moment`."""
+    return any(op.gate == gate for op in moment)
+
+
+@transformer_api.transformer
+class DepolerizingNoiseTransformer:
+    """Add local depolarizing noise after two-qubit gates in a specified circuit. More specifically,
+    with probability p, append a random non-identity two-qubit Pauli operator after each specified
+    two-qubit gate.
+
+    Attrs:
+        p: The probability with which to add noise.
+        target_gate: Add depolarizing nose after this type of gate
+    """
+
+    def __init__(
+        self, p: float | Mapping[tuple[ops.Qid, ops.Qid], float], target_gate: ops.Gate = ops.CZ
+    ):
+        """Initialize the depolarizing noise transformer with some depolarizing probability and
+        target gate.
+
+        Args:
+            p: The depolarizing probability, either a single float or a mapping from pairs of qubits
+               to floats.
+           target_gate: The gate after which to add depolarizing noise.
+
+        Raises:
+            TypeError: If `p` is not either be a float or a mapping from sorted qubit pairs to
+                       floats.
+        """
+
+        if not isinstance(p, (Mapping, float)):
+            raise TypeError(  # pragma: no cover
+                "p must either be a float or a mapping from"  # pragma: no cover
+                + "sorted qubit pairs to floats"  # pragma: no cover
+            )  # pragma: no cover
+        self.p = p
```

Comment:
you can resolve `p` here and then use it in __call__ as `self.p_func(pair)`
```suggestion
        self.p_func = lambda _: p if isinstance(p, (int, float)) else lambda pair: p.get(pair, 0)
```